x/projects/keeper/project_test.go

```diff
@@ -671,6 +671,68 @@ func TestChargeComputeUnits(t *testing.T) {
 	require.Equal(t, uint64(0), proj.UsedCu)
 }
 
+func TestAddAfterDelKeys(t *testing.T) {
+	ts := newTester(t)
+	ts.SetupAccounts(1, 0, 2) // 1 sub, 0 adm, 2 dev
+
+	_, sub1Addr := ts.Account("sub1")
+	_, dev1Addr := ts.Account("dev1")
+	_, dev2Addr := ts.Account("dev2")
+
+	plan := ts.Plan("mock")
+
+	projectData1 := types.ProjectData{
+		Name:        "mockname1",
+		Enabled:     true,
+		ProjectKeys: []types.ProjectKey{},
+		Policy:      &plan.PlanPolicy,
+	}
+	err := ts.Keepers.Projects.CreateProject(ts.Ctx, sub1Addr, projectData1, plan)
+	require.Nil(t, err)
+
+	projectData2 := types.ProjectData{
+		Name:        "mockname2",
+		Enabled:     true,
+		ProjectKeys: []types.ProjectKey{types.ProjectDeveloperKey(dev2Addr)},
+		Policy:      &plan.PlanPolicy,
+	}
+	err = ts.Keepers.Projects.CreateProject(ts.Ctx, sub1Addr, projectData2, plan)
+	require.Nil(t, err)
+
+	ts.AdvanceBlock()
+
+	projectID1 := types.ProjectIndex(sub1Addr, projectData1.Name)
+	projectID2 := types.ProjectIndex(sub1Addr, projectData2.Name)
+
+	// add key
+	err = ts.addProjectKeys(projectID1, sub1Addr, types.ProjectDeveloperKey(dev1Addr))
+	require.Nil(t, err)
+	require.True(t, ts.isKeyInProject(projectID1, dev1Addr, types.ProjectKey_DEVELOPER))
+	require.True(t, ts.isKeyInProject(projectID2, dev2Addr, types.ProjectKey_DEVELOPER))
+
+	// add same key to other project - should fail
+	err = ts.addProjectKeys(projectID2, sub1Addr, types.ProjectDeveloperKey(dev1Addr))
+	require.NotNil(t, err)
+
+	// del key
+	err = ts.delProjectKeys(projectID1, sub1Addr, types.ProjectDeveloperKey(dev1Addr))
+	require.Nil(t, err)
+
+	// del takes effect in next epoch
+	require.True(t, ts.isKeyInProject(projectID1, dev1Addr, types.ProjectKey_DEVELOPER))
```

Comment:
adding the key should work even before we advance the epoch and should apply next then, lets test it here

Reply:
in the current semantics, adding the same entry again is only possible after the entry has reached its `deleteAt` block. this matches the current semantics of adding projects (or keys) which takes place now (or at the beginning of this epoch) - so cannot complete successfully while a delete is pending.

we can extend this to allow adding the same entry in the future (including at/on the current `deleteAt`) if necessary. but that is more subtle so left for a separate PR.